Add inlined isAssignableFrom test on X

matthewhall2 opened the pull request:
Adds inlined isAssignableFrom test, skipping the slow helper in most cases. This also adds logic to check if we know the type of the class (whether its an interface, array class, or other) at compile time and will only generate code for that specific case, skipping tests that will fail. For example, the superclass test will always fail for interfaces. So, if we know we have an interface at compile time, we only generate the itable test (an inline itable walk).

Commons out superclass test and itable walk from `generateInlinedCheckCastForDynamicCastClass` since isAssignableFrom uses the same tests

hzongaro reviewed (commented):
I think the changes look functionally correct.  Just a couple of mainly cosmetic comments.

May I also ask you to change "isAssignableFrom" to "Class.isAssignableFrom" in the commit title and body, just for clarity?

One additional question:  are there cases where the JIT can determine at compile-time whether the Class objects involved are array classes or interface classes?  I'm just wondering how often the opportunity to eliminate any of the run-time tests arises in practice.

matthewhall2 committed:
Add inlined Class.isAssignableFrom test on X

Adds inlined Class.isAssignableFrom test, skipping the slow helper in most
cases. This also adds logic to check for the difference cases (interface
class, array class, anything else) to skip tests if we have the class
info at compile time.

Signed-off-by: Matthew Hall <matthew.hall3@outlook.com>

matthewhall2 commented:
@hzongaro as said this is now only for the case where the castClass is unknown at compile time, so the previous fast-paths are not needed.
